web/src/js/ducks/websocket.js

```diff
@@ -1,30 +1,103 @@
+import * as eventLogActions from './eventLog'
+import * as flowActions from './flows'
+
+const CONNECT = 'WEBSOCKET_CONNECT'
 const CONNECTED = 'WEBSOCKET_CONNECTED'
+const DISCONNECT = 'WEBSOCKET_DISCONNECT'
 const DISCONNECTED = 'WEBSOCKET_DISCONNECTED'
+const ERROR = 'WEBSOCKET_ERROR'
+const MESSAGE = 'WEBSOCKET_MESSAGE'
 
+/* we may want to have an error message attribute here at some point */
+const defaultState = { connected: false, socket: null }
 
-const defaultState = {
-    connected: false,
-    /* we may want to have an error message attribute here at some point */
-}
 export default function reducer(state = defaultState, action) {
     switch (action.type) {
+
+        case CONNECT:
+            return { ...state, socket: action.socket }
+
         case CONNECTED:
-            return {
-                connected: true
-            }
+            return { ...state, connected: true }
+
+        case DISCONNECT:
+            return { ...state, connected: false }
+
         case DISCONNECTED:
-            return {
-                connected: false
-            }
+            return { ...state, socket: null }
+
         default:
             return state
     }
 }
 
+export function connect() {
+    return dispatch => {
+        const socket = new WebSocket(location.origin.replace('http', 'ws') + '/updates')
+
+        // @todo remove this
+        window.ws = socket
+
+        socket.addEventListener('open', () => dispatch(onConnect()))
+        socket.addEventListener('close', () => dispatch(onDisconnect()))
+        socket.addEventListener('message', msg => dispatch(onMessage(msg)))
+        socket.addEventListener('error', error => dispatch(onError(error)))
+
+        dispatch({ type: CONNECT, socket })
+
+        return socket
+    }
+}
+
+export function disconnect() {
+    return { type: DISCONNECT }
+}
+
+export function onConnect() {
+    // workaround to make sure that our state is already available.
+    return dispatch => {
+        dispatch({ type: CONNECTED })
+        dispatch(flowActions.fetchFlows()).then(() => ConnectionActions.open())
+    }
+}
+
+export function onDisconnect() {
+    return dispatch => {
+        ConnectionActions.close()
+        dispatch(eventLogActions.addLogEntry('WebSocket connection closed.'))
+        dispatch({ type: DISCONNECTED })
+    }
+}
+
+export function onMessage(msg) {
+    return dispatch => {
+        const data = JSON.parse(msg.data)
+
+        AppDispatcher.dispatchServerAction(data)
 
-export function connected() {
-    return {type: CONNECTED}
+        switch (data.type) {
+
+            case eventLogActions.UPDATE_LOG:
+                return dispatch(eventLogActions.updateLogEntries(data))
+
+            case flowActions.UPDATE_FLOWS:
+                return dispatch(flowActions.updateFlows(data))
+
+            default:
+                console.warn('unknown message', data)
+        }
+
+        dispatch({ type: MESSAGE, msg })
```

Comment:
The problem I see with this generic approach is that the server may send messages which do no modify any redux state, but we won't notice that, as the action just passes by everything and does not cause any change.
Is there anything you did not like about the `flowActions.updateFlows` etc approach, which provided one dedicated actionCreator for each type of websocket message?